wandb/apis/public.py

```diff
@@ -2377,7 +2382,98 @@ def __repr__(self):
         return "<ArtifactCollection {} ({})>".format(self.name, self.type)
 
 
+def _determine_artifact_root(source_path, target_path):
+    """Helper function to determine the artifact root of `target_path` by comparing to
+    an existing artifact asset in `source`path`. This is used in reference artifact resolution"""
+    abs_source_path = os.path.abspath(source_path)
+    abs_target_path = os.path.abspath(target_path)
+
+    # Break the source path into parts
+    head = abs_source_path
+    source_path_parts = []
+    while head != "" and head != "/":
+        head, tail = os.path.split(head)
+        source_path_parts.insert(0, tail)
```

Comment:
It's not very pythonic to insert at the head in a loop (inserting at the head is O(n) so the loop is O(n^2)). Instead you could build it backward then reverse it. But just use os.path.split

Reply:
fixed in next commit